bpo-31324: Optimize support._match_test()

vstinner opened the pull request:
* Remove support.match_tests global variable. It is replaced with a
  new support.set_match_tests() function, so _match_test() doesn't
  have to check each time if patterns were modified.
* Rewrite _match_test() with a new _MatchTests object which uses
  different code paths depending on the kind of patterns.

Co-Authored-By: Serhiy Storchaka <storchaka@gmail.com>

<!-- issue-number: bpo-31324 -->
https://bugs.python.org/issue31324
<!-- /issue-number -->

vstinner committed:
bpo-31324: Optimize support.match_test()

* Rename support._match_test() to support.match_test(): make it
  public
* Remove support.match_tests global variable. It is replaced with a
  new support.set_match_tests() function, so match_test() doesn't
  have to check each time if patterns were modified.
* Rewrite match_test(): use different code paths depending on the
  kind of patterns for best performances.

Co-Authored-By: Serhiy Storchaka <storchaka@gmail.com>

vstinner commented:
`Lib/test/support/__init__.py`:

```
-match_tests = None
```

Only Lib/test/libregrtest/ was using support.match_tests, and my PR updates Lib/test/libregrtest/. I don't think that support.mtach_tests is used outside CPython code base. test.support is mostly used internally by the CPython test suite.

serhiy-storchaka commented on Lib/test/support/__init__.py:
```diff
@@ -1900,21 +1899,48 @@ def _run_suite(suite):
+
+
+def _match_test(test):
+    global _match_test_obj
```
Not needed.

vstinner replied:
Fixed ("global" removed)

serhiy-storchaka commented on Lib/test/support/__init__.py:
```diff
@@ -1900,21 +1899,48 @@ def _run_suite(suite):
 
-def _match_test(test):
-    global match_tests
+class _MatchTests:
```
I think the class is not needed. Just use a function which returns a function.

vstinner replied:
Ok, done.

serhiy-storchaka commented on Lib/test/support/__init__.py:
```diff
@@ -1900,21 +1899,48 @@ def _run_suite(suite):
+            # Simple case: all patterns are full test identifier.
+            # The test.bisect utility uses such patterns.
+            self._pattern_set = set(patterns)
+            self.match = self._full_match
```
Just return `set(patterns).__contains__`.

vstinner replied:
Done

serhiy-storchaka commented on Lib/test/support/__init__.py:
```diff
@@ -1900,21 +1899,48 @@ def _run_suite(suite):
+        return any(map(self.regex_match, test_id.split(".")))
+
+    def _is_full_match(self, pattern):
+        return ('.' in pattern) and ('*' not in pattern)
```
What about '?' and '['?

vstinner replied:
Fixed

serhiy-storchaka commented on Lib/test/support/__init__.py:
```diff
@@ -1900,21 +1899,48 @@ def _run_suite(suite):
-    global match_tests
+class _MatchTests:
+    def __init__(self, patterns):
+        if all(self._is_full_match(pattern) for pattern in patterns):
```
Could be `all(map(self._is_full_match, patterns))`.

vstinner commented:
@serhiy-storchaka: I addressed your comments, I added unit tests, I chose to make support.match_test() public since it's used by regrtest, and I rebased my change.

serhiy-storchaka commented on Lib/test/support/__init__.py:
```diff
@@ -1900,21 +1899,54 @@ def _run_suite(suite):
+    #
+    # Reject patterns which contain fnmatch patterns: '*', '?', '[...]'
+    # or '[!...]'. For example, reject 'test_access*'.
+    return ('.' in pattern) and all(char not in '?*[]' for char in pattern)
```
I'm wondering whether using a regular expression would be faster.
```python
return ('.' in pattern) and not re.search(r'[?*\[\]]', pattern)
```

vstinner replied:
I think that (char not in '?*[]') is more readable, and this function doesn't seem to be a performance bottleneck, there is no need to optimize it. Moreover, I just added a cache to not recompile patterns if set_match_tests() is called twice with the same patterns.

It seems like the compilation to a giant regex is slow, but I don't think that we can optimize this much, and again, I don't think that it's worth it. My main usage of many patterns is when I use test.bisect, but test.bisect only uses full test identifiers and so should use set().__contains__.

serhiy-storchaka commented on Lib/test/support/__init__.py:
```diff
@@ -1900,21 +1899,54 @@ def _run_suite(suite):
+
+    if not patterns:
+        _match_test_func = None
+    elif all(_is_full_match_test(pattern) for pattern in patterns):
```
Could be `all(map(_is_full_match_test, patterns))`.

vstinner replied:
Right, but I prefer list comprehensions. IMHO it's more readable.

serhiy-storchaka replied:
`map()` is already used twice few lines below.

vstinner replied:
Oh right, that's the copied I copied from your PR :-D You won, I updated my PR to address your latest review.

serhiy-storchaka commented on Lib/test/test_support.py:
```diff
@@ -483,6 +483,48 @@ def test_optim_args_from_interpreter_flags(self):
+        test_access = Test('test.test_os.FileTests.test_access')
+        test_chdir = Test('test.test_os.Win32ErrorTests.test_chdir')
+
+        old_match = support._match_test_func
```
Could use
```python
with support.swap_attr(support, '_match_test_func', None):
```

vstinner replied:
Done

serhiy-storchaka reviewed (commented):
Added few suggestions.

And please add tests with a pattern containing `'.'` and metacharacters and with multiple patterns.

vstinner committed:
set_match_tests() does nothing if patterns don't change

vstinner committed:
Add more tests

Use also support.swap_attr().

vstinner commented:
> And please add tests with a pattern containing '.' and metacharacters and with multiple patterns.

Done

vstinner committed:
Address Serhiy's comments

serhiy-storchaka commented on Lib/test/support/__init__.py:
```diff
@@ -1913,12 +1913,13 @@ def match_test(test):
     # Reject patterns which contain fnmatch patterns: '*', '?', '[...]'
     # or '[!...]'. For example, reject 'test_access*'.
-    return ('.' in pattern) and all(char not in '?*[]' for char in pattern)
+    return ('.' in pattern) and (not re.search(r'[?*\[\]]', pattern))
```
Are parenthesis around `not ...` add readability?

serhiy-storchaka commented on Lib/test/support/__init__.py:
```diff
@@ -1900,21 +1899,64 @@ def _run_suite(suite):
+
+        func = match_test_regex
+
+    _match_test_patterns = patterns
```
Shouldn't we make a copy? Or just ignore the case of modifying patterns after calling  `set_match_tests()`? This is a private API at end.

vstinner replied:
Right, fixed.

vstinner committed:
set_match_tests() copies patterns

vstinner commented:
Thank you for the reviews @serhiy-storchaka!
